[OCPCLOUD-1739] Add full veritcal scaling rolling update periodic test

JoelSpeed opened the pull request:
This PR does a number of things working towards the goal of a vertical scaling, rolling update test.

It's likely best reviewed commit wise:
* Updating ginkgo and gomega to the latest versions
* Setting fail-fast on the E2E runner
  * Fail fast means that if one test fails, we don't run the others, this is what we use for other E2Es
* Wait for cluster operators to stabilise before running tests
  * This checks that all of the cluster operators are available, are not degraded and not progressing, ie the cluster is stable, we run should this before each of the tests that interfere with cluster stability
* Add custom formatter for conditions
  * While debugging I realised it's hard to work out from a list of 35 operators which ones failed, this formatter extracts only the failed operators and returns those, makes it easier to process a failure output
* Add pre-step to ensure CPMS is present and active
  * On platforms that don't have automatic support, we will want to extend this to create a CPMS (eg Azure or vSphere)
  * This checks that the CPMS is present and activates it if it's not, this should allow other previous tests to, for example, leave an inactive CPMS lying around
* Add IncreaseProviderSpecInstanceSize
  * This will need to be implemented for each platform we want to test on
  * It basically works out how to make the instance bigger and updates the spec
  * This should be able to be used for the other checks that use a single machine too
* Add async utils
  * I wanted to be able to run lots of checks in parallel and have them report back into the same `It`
  * This adds an async wrapper than handles context cancellation and wait group addition for an aysnc function
  * Also adds RunCheckUntil which allows you to check a condition `Consistently` until some other condition `Eventually` passes, useful for example for check the Machine does not get deleted until it's replacement is up and running
*   Add checks for machine replacements
  * Checks that the correct/expected index is the one that is being replaced
  * Checks that the replacement of the Machine happens smoothly, in the order expected for a rolling update
* Add periodic to check RollingUpdate
  * Ties all of the previous work together into a periodic test that checks the rolling update procedure 

I've manually tested this on a cluster and seen both failures and success (as I was expecting) by playing around a bit.

I've taken a lot of care to make sure that the failures are reported immediately and that we don't continue running a test when we know it's already failed, this should allow quicker feedback from the E2Es

lobziik commented on test/e2e/periodic_test.go:
```diff
@@ -17,11 +17,32 @@ limitations under the License.
+				periodic.IncreaseControlPlaneMachineSetInstanceSize(testFramework)
+			})
+
+			periodic.ItShouldPerformARollingUpdate(testFramework)
```
First couple of reads i didn't recognise this thing as a test itself.

Honestly, this pattern with moving tests into separate packages and wrapping 'It' with another function feels a bit tangled and not so convenient to review.

Perhaps we could make this more flat?

JoelSpeed replied:
So this is a common pattern that is recommended in the Ginkgo documentation to make test suites composable. The suite itself is composed of tests, we don't really care about the details of the tests at this level but instead want to see what tests there are. That's why I'm keeping the name as `ItShouldPerformARollingUpdate` for example.

What I don't want to do is move the bulk of the test into this function because then you lose that composability, and, we have to move all of the helpers out too. What we have here means that we can keep the test and all of it's helpers neatly together, and, if we wanted to, we could re-use them in an integration suite as well (we'd have to mock the machine behaviour but otherwise it would work).

Do you think it would be easier if it were as below?

```
It("performs a rolling update", func() {
  periodic.VerifyRollingUpdate(testFramework)
})
```

lobziik replied:
>Do you think it would be easier if it were as below?
```
It("performs a rolling update", func() {
  periodic.VerifyRollingUpdate(testFramework)
})
```

At first glance - yes. I would prefer this.

JoelSpeed replied:
Let's get some opinions, I still prefer the other way but happy to yield if most think it's better that way around

damdo replied:
If the change we are discussing is having an internal vs external It wrapping, I don't have strong opinions here, as I see it as only a cosmetic opinion. I'll let you two decide :)

lobziik commented on test/e2e/periodic/periodic.go:
```diff
@@ -0,0 +1,183 @@
+// IncreaseControlPlaneMachineSetInstanceSize increases the instance size of the control plane machine set.
+// This should trigger the control plane machine set to update the machines based on the
+// update strategy.
+func IncreaseControlPlaneMachineSetInstanceSize(testFramework framework.Framework, args ...interface{}) machinev1beta1.ProviderSpec {
```
Is it necessary to have `args ...interface{}` here?

This args uses only as an `Eventually` arguments, maybe we could choose a more descriptive name for this function, which explains what this `args` is about. Or chose another name for this argument itself

JoelSpeed replied:
`args ...interface{}` comes from the gomega framework, it's how they create variadic arguments for the timeouts or contexts for the async assertions.

So normally you can do `Eventually(func)` or `Eventually(func, timeout)` or `Eventually(func, timeout, interval)` or `Eventually(func, ctx)`, I wanted to make it so that we could configure that at the parent test suite level as well so mimicked their pattern, do you think that's not worth doing for now and we should just set the timeouts based on the work in the individual functions?

For the control plane to stabilise one, I think it's important to have this variable as we may want to have a shorter check in the pre-step then we do at the end of say a rolling update, but I'm less concerned about it here for example

lobziik replied:
I believe, i did get the purpose of `args` after I read through the function itself but it was quite cryptic when i look on signature.

In other similar places there are some hints in function name. I'm totally fine to keep this pattern, just want to make function signature more clear.

JoelSpeed replied:
> In other similar places there are some hints in function names. I'm totally fine to keep this pattern, just want to make function signature more clear.

Do you have an example?

Calling them `gomegaArgs` works for me if that helps clarify it

lobziik replied:
>Do you have an example?

I thought about `EventuallyClusterOperatorsShouldStabilise` when I was writing that. 

Having `gomegaArgs` everywhere even better, thanks!

lobziik reviewed (commented):
Looks cool! I certainly do like async utils set, think it would be useful in other MAPI related tests.

Left some comments below.

I would like to get back to this tomorrow, want to give it another fresh look.

JoelSpeed commented on test/e2e/periodic/periodic.go:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	getCPMSArgs := append([]interface{}{komega.Get(cpms)}, args...)
+	Eventually(getCPMSArgs...).Should(Succeed(), "control plane machine set should exist")
```
Let me look at this, I didn't originally move it out because the code to fetch the CPMS is pretty small, but I can look.

In general I'm trying to keep ginkgo/gomega out of the `framework`, so it could create a function that gets the CPMS and returns an error, but then getting access to the CPMS object is a bit of hard work

lobziik replied:
`NewEmptyControlPlaneMachineSet` works for me, thanks

JoelSpeed commented:
Ran this successfully, output is as below, the output is a little verbose but honestly, that's probably not a bad thing in this case
```
INKGO_EXTRA_ARGS=--label-filter=Periodic ./hack/e2e.sh
go run ./hack/../vendor/github.com/onsi/ginkgo/v2/ginkgo -r -v --fail-fast --trace --label-filter=Periodic ./test/e2e
I1102 15:32:23.287660   27431 request.go:682] Waited for 1.049324209s due to client-side throttling, not priority and fairness, request: GET:https://api.jspeed-test-2.devcluster.openshift.com:6443/apis/flowcontrol.apiserver.k8s.io/v1beta2?timeout=32s
Running Suite: E2E Suite - /Users/jspeed/go/src/github.com/openshift/cluster-control-plane-machine-set-operator/test/e2e
========================================================================================================================
Random Seed: 1667399539

Will run 2 of 2 specs
------------------------------
[BeforeSuite] 
/Users/jspeed/go/src/github.com/openshift/cluster-control-plane-machine-set-operator/test/e2e/suite_test.go:50
------------------------------
[BeforeSuite] PASSED [0.000 seconds]
[BeforeSuite] 
/Users/jspeed/go/src/github.com/openshift/cluster-control-plane-machine-set-operator/test/e2e/suite_test.go:50
------------------------------
ControlPlaneMachineSet Operator With an active ControlPlaneMachineSet and the instance type is changed
  should perform a rolling update [Periodic]
  /Users/jspeed/go/src/github.com/openshift/cluster-control-plane-machine-set-operator/test/e2e/periodic_test.go:44
STEP: Waiting for the cluster operators to stabilise 11/02/22 15:32:24.696
STEP: Checking the control plane machine set exists 11/02/22 15:32:25.095
STEP: Checking the control plane machine set is active 11/02/22 15:32:25.226
STEP: Increasing the control plane machine set instance size 11/02/22 15:32:25.486
STEP: Waiting for the index 0 to be replaced 11/02/22 15:32:25.889
STEP: Checking that other indexes (not 0) do not have 2 replicas 11/02/22 15:32:25.889
STEP: Checking the number of control plane machines never goes above 4 replicas 11/02/22 15:32:25.891
STEP: Waiting for the updated replicas to equal desired replicas 11/02/22 15:32:25.891
STEP: Checking that index 0 has 2 replicas 11/02/22 15:32:25.891
STEP: Correct index is being replaced 11/02/22 15:32:26.026
STEP: Index 0 replacement created 11/02/22 15:32:26.026
STEP: Checking the replacement machine for index 0 11/02/22 15:32:26.026
STEP: Checking the replacement machine name 11/02/22 15:32:26.158
STEP: Replacement machine name is correct 11/02/22 15:32:26.158
STEP: Checking that the old machine is not deleted until the new machine is Running 11/02/22 15:32:26.158
STEP: Checking that the new machine is Running 11/02/22 15:32:26.159
STEP: Replacement machine is Running 11/02/22 15:35:10.805
STEP: Checking that the old machine is marked for deletion 11/02/22 15:35:10.806
STEP: Checking that the old machine is removed 11/02/22 15:35:10.934
STEP: Rollout of index 0 complete 11/02/22 15:40:29.512
STEP: Replacement for index 0 is complete 11/02/22 15:40:29.513
STEP: Waiting for the index 1 to be replaced 11/02/22 15:40:29.513
STEP: Checking that other indexes (not 1) do not have 2 replicas 11/02/22 15:40:29.513
STEP: Checking that index 1 has 2 replicas 11/02/22 15:40:29.513
STEP: Correct index is being replaced 11/02/22 15:40:29.894
STEP: Index 1 replacement created 11/02/22 15:40:29.895
STEP: Checking the replacement machine for index 1 11/02/22 15:40:29.895
STEP: Checking the replacement machine name 11/02/22 15:40:30.029
STEP: Replacement machine name is correct 11/02/22 15:40:30.029
STEP: Checking that the old machine is not deleted until the new machine is Running 11/02/22 15:40:30.029
STEP: Checking that the new machine is Running 11/02/22 15:40:30.029
STEP: Replacement machine is Running 11/02/22 15:43:09.455
STEP: Checking that the old machine is marked for deletion 11/02/22 15:43:09.456
STEP: Checking that the old machine is removed 11/02/22 15:43:09.587
STEP: Rollout of index 1 complete 11/02/22 15:55:32.641
STEP: Replacement for index 1 is complete 11/02/22 15:55:32.641
STEP: Waiting for the index 2 to be replaced 11/02/22 15:55:32.641
STEP: Checking that other indexes (not 2) do not have 2 replicas 11/02/22 15:55:32.642
STEP: Checking that index 2 has 2 replicas 11/02/22 15:55:32.643
STEP: Correct index is being replaced 11/02/22 15:55:33.01
STEP: Index 2 replacement created 11/02/22 15:55:33.011
STEP: Checking the replacement machine for index 2 11/02/22 15:55:33.011
STEP: Checking the replacement machine name 11/02/22 15:55:33.157
STEP: Replacement machine name is correct 11/02/22 15:55:33.157
STEP: Checking that the old machine is not deleted until the new machine is Running 11/02/22 15:55:33.157
STEP: Checking that the new machine is Running 11/02/22 15:55:33.157
STEP: Updated replicas is now equal to desired replicas 11/02/22 15:57:51.401
STEP: Waiting for the replicas to equal desired replicas 11/02/22 15:57:51.402
STEP: Replacement machine is Running 11/02/22 15:57:51.953
STEP: Checking that the old machine is marked for deletion 11/02/22 15:57:51.953
STEP: Checking that the old machine is removed 11/02/22 15:57:52.077
STEP: Replicas is now equal to desired replicas 11/02/22 16:08:43.682
STEP: Rollout of index 2 complete 11/02/22 16:08:44.189
STEP: Replacement for index 2 is complete 11/02/22 16:08:44.189
STEP: Control plane machine completed successfully 11/02/22 16:08:44.189
STEP: Waiting for the cluster to stabilise after the rollout 11/02/22 16:08:44.189
STEP: Waiting for the cluster operators to stabilise 11/02/22 16:08:44.189
STEP: Cluster stabilised after the rollout 11/02/22 16:23:38.405
------------------------------
• [SLOW TEST] [2735.389 seconds]
ControlPlaneMachineSet Operator [Periodic]
/Users/jspeed/go/src/github.com/openshift/cluster-control-plane-machine-set-operator/test/e2e/periodic_test.go:29
  With an active ControlPlaneMachineSet
  /Users/jspeed/go/src/github.com/openshift/cluster-control-plane-machine-set-operator/test/e2e/periodic_test.go:34
    and the instance type is changed
    /Users/jspeed/go/src/github.com/openshift/cluster-control-plane-machine-set-operator/test/e2e/periodic_test.go:39
      should perform a rolling update
      /Users/jspeed/go/src/github.com/openshift/cluster-control-plane-machine-set-operator/test/e2e/periodic_test.go:44

  Begin Captured GinkgoWriter Output >>
    STEP: Waiting for the cluster operators to stabilise 11/02/22 15:32:24.696
    STEP: Checking the control plane machine set exists 11/02/22 15:32:25.095
    STEP: Checking the control plane machine set is active 11/02/22 15:32:25.226
    STEP: Increasing the control plane machine set instance size 11/02/22 15:32:25.486
    STEP: Waiting for the index 0 to be replaced 11/02/22 15:32:25.889
    STEP: Checking that other indexes (not 0) do not have 2 replicas 11/02/22 15:32:25.889
    STEP: Checking the number of control plane machines never goes above 4 replicas 11/02/22 15:32:25.891
    STEP: Waiting for the updated replicas to equal desired replicas 11/02/22 15:32:25.891
    STEP: Checking that index 0 has 2 replicas 11/02/22 15:32:25.891
    STEP: Correct index is being replaced 11/02/22 15:32:26.026
    STEP: Index 0 replacement created 11/02/22 15:32:26.026
    STEP: Checking the replacement machine for index 0 11/02/22 15:32:26.026
    STEP: Checking the replacement machine name 11/02/22 15:32:26.158
    STEP: Replacement machine name is correct 11/02/22 15:32:26.158
    STEP: Checking that the old machine is not deleted until the new machine is Running 11/02/22 15:32:26.158
    STEP: Checking that the new machine is Running 11/02/22 15:32:26.159
    STEP: Replacement machine is Running 11/02/22 15:35:10.805
    STEP: Checking that the old machine is marked for deletion 11/02/22 15:35:10.806
    STEP: Checking that the old machine is removed 11/02/22 15:35:10.934
    STEP: Rollout of index 0 complete 11/02/22 15:40:29.512
    STEP: Replacement for index 0 is complete 11/02/22 15:40:29.513
    STEP: Waiting for the index 1 to be replaced 11/02/22 15:40:29.513
    STEP: Checking that other indexes (not 1) do not have 2 replicas 11/02/22 15:40:29.513
    STEP: Checking that index 1 has 2 replicas 11/02/22 15:40:29.513
    STEP: Correct index is being replaced 11/02/22 15:40:29.894
    STEP: Index 1 replacement created 11/02/22 15:40:29.895
    STEP: Checking the replacement machine for index 1 11/02/22 15:40:29.895
    STEP: Checking the replacement machine name 11/02/22 15:40:30.029
    STEP: Replacement machine name is correct 11/02/22 15:40:30.029
    STEP: Checking that the old machine is not deleted until the new machine is Running 11/02/22 15:40:30.029
    STEP: Checking that the new machine is Running 11/02/22 15:40:30.029
    STEP: Replacement machine is Running 11/02/22 15:43:09.455
    STEP: Checking that the old machine is marked for deletion 11/02/22 15:43:09.456
    STEP: Checking that the old machine is removed 11/02/22 15:43:09.587
    STEP: Rollout of index 1 complete 11/02/22 15:55:32.641
    STEP: Replacement for index 1 is complete 11/02/22 15:55:32.641
    STEP: Waiting for the index 2 to be replaced 11/02/22 15:55:32.641
    STEP: Checking that other indexes (not 2) do not have 2 replicas 11/02/22 15:55:32.642
    STEP: Checking that index 2 has 2 replicas 11/02/22 15:55:32.643
    STEP: Correct index is being replaced 11/02/22 15:55:33.01
    STEP: Index 2 replacement created 11/02/22 15:55:33.011
    STEP: Checking the replacement machine for index 2 11/02/22 15:55:33.011
    STEP: Checking the replacement machine name 11/02/22 15:55:33.157
    STEP: Replacement machine name is correct 11/02/22 15:55:33.157
    STEP: Checking that the old machine is not deleted until the new machine is Running 11/02/22 15:55:33.157
    STEP: Checking that the new machine is Running 11/02/22 15:55:33.157
    STEP: Updated replicas is now equal to desired replicas 11/02/22 15:57:51.401
    STEP: Waiting for the replicas to equal desired replicas 11/02/22 15:57:51.402
    STEP: Replacement machine is Running 11/02/22 15:57:51.953
    STEP: Checking that the old machine is marked for deletion 11/02/22 15:57:51.953
    STEP: Checking that the old machine is removed 11/02/22 15:57:52.077
    STEP: Replicas is now equal to desired replicas 11/02/22 16:08:43.682
    STEP: Rollout of index 2 complete 11/02/22 16:08:44.189
    STEP: Replacement for index 2 is complete 11/02/22 16:08:44.189
    STEP: Control plane machine completed successfully 11/02/22 16:08:44.189
    STEP: Waiting for the cluster to stabilise after the rollout 11/02/22 16:08:44.189
    STEP: Waiting for the cluster operators to stabilise 11/02/22 16:08:44.189
    STEP: Cluster stabilised after the rollout 11/02/22 16:23:38.405
  << End Captured GinkgoWriter Output
------------------------------
ControlPlaneMachineSet Operator on a fully supported cluster upon install
  should have an active control plane machine set [Periodic, PreSubmit]
  /Users/jspeed/go/src/github.com/openshift/cluster-control-plane-machine-set-operator/test/e2e/common_test.go:42
STEP: Waiting for the cluster operators to stabilise 11/02/22 16:23:38.42
------------------------------
• [0.405 seconds]
ControlPlaneMachineSet Operator [Periodic, PreSubmit]
/Users/jspeed/go/src/github.com/openshift/cluster-control-plane-machine-set-operator/test/e2e/common_test.go:29
  on a fully supported cluster
  /Users/jspeed/go/src/github.com/openshift/cluster-control-plane-machine-set-operator/test/e2e/common_test.go:34
    upon install
    /Users/jspeed/go/src/github.com/openshift/cluster-control-plane-machine-set-operator/test/e2e/common_test.go:41
      should have an active control plane machine set
      /Users/jspeed/go/src/github.com/openshift/cluster-control-plane-machine-set-operator/test/e2e/common_test.go:42

  Begin Captured GinkgoWriter Output >>
    STEP: Waiting for the cluster operators to stabilise 11/02/22 16:23:38.42
  << End Captured GinkgoWriter Output
------------------------------

Ran 2 of 2 Specs in 2735.809 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 1 suite in 45m41.507534125s
```

JoelSpeed commented:
/test e2e-aws-periodic-pre

JoelSpeed commented:
/test e2e-aws-periodic-pre

JoelSpeed commented:
/test e2e-aws-periodic-pre

JoelSpeed commented:
/test unit

damdo commented on test/e2e/periodic/periodic.go:
```diff
@@ -0,0 +1,177 @@
+		cpms := &machinev1.ControlPlaneMachineSet{}
+		Expect(k8sClient.Get(ctx, framework.ControlPlaneMachineSetKey(), cpms)).To(Succeed(), "control plane machine set should exist")
+
+		// We give the rollout an hour to complete.
+		// We pass this to Eventually and Consistently assertions to ensure that they check
+		// until they pass or until the timeout is reached.
+		rolloutCtx, cancel := context.WithTimeout(testFramework.GetContext(), 1*time.Hour)
+		defer cancel()
+
+		wg := &sync.WaitGroup{}
```
Is this a comfortable timeout? Otherwise we would race against the speed of the cloud provider rather than verifying the rollout completion.

JoelSpeed replied:
At the moment, no, the timeout isn't configured and it's pinned to 1 hour. I think given our intention of testing this on AWS, Azure, and in the future GCP and vSphere, 1 hour is probably enough for all of those, but we could introduce into the framework a concept of platform specific timeouts.

I think for now this should be ok and we can implement something configurable if we need to later

damdo commented on test/e2e/periodic/periodic.go:
```diff
@@ -0,0 +1,177 @@
+
+		wg := &sync.WaitGroup{}
+
+		framework.Async(wg, cancel, func() bool {
+			By("Checking the number of control plane machines never goes above 4 replicas")
+			controlPlaneMachineSelector := runtimeclient.MatchingLabels(framework.ControlPlaneMachineSetSelectorLabels())
+
+			return Consistently(komega.ObjectList(&machinev1beta1.MachineList{}, controlPlaneMachineSelector), rolloutCtx).Should(HaveField("Items", SatisfyAny(
+				HaveLen(3),
+				HaveLen(4),
+			)), "control plane machines should never go above 4 replicas, or below 3 replicas")
+		})
```
Should we mention the fact that we are checking the (for now) hard coded surge?

JoelSpeed replied:
Yes, I'll add a comment that explains this

damdo commented on test/e2e/periodic/periodic.go:
```diff
@@ -0,0 +1,177 @@
+// waitForReplicas waits for the control plane machine set to have the desired number of replicas.
+// It first waits for the updated replicas to equal the desired number, and then waits for the final replica
+// count to equal the desired number.
+func waitForReplicas(ctx context.Context, cpms *machinev1.ControlPlaneMachineSet) bool {
```
```suggestion
func waitForDesiredReplicas(ctx context.Context, cpms *machinev1.ControlPlaneMachineSet) bool {
```
Would this be more inline with what it does then?

damdo commented on test/e2e/periodic/periodic.go:
```diff
@@ -0,0 +1,183 @@
+
+// checkRolloutProgress monitors the progress of each index in the rollout in turn.
+func checkRolloutProgress(testFramework framework.Framework, ctx context.Context) bool {
+	if ok := checkRolloutForIndex(testFramework, ctx, 0); !ok {
+		return false
+	}
+
+	if ok := checkRolloutForIndex(testFramework, ctx, 1); !ok {
+		return false
+	}
+
+	if ok := checkRolloutForIndex(testFramework, ctx, 2); !ok {
+		return false
+	}
```
Are we guaranteed to always have these particular indexes in the e2e?

JoelSpeed replied:
In E2E yes, these are the indexes the installer creates by default and we expect a clean cluster. If the indexes the installer creates change, this will need to be updated, but for now this test assumes a 0, 1, 2 configuration and I think that's ok

damdo replied:
Yes, I agree, was just thinking for the worst to see if there was anything we missed.

damdo reviewed (commented):
Great!
Left some comments but looks good as a whole.

JoelSpeed commented on test/e2e/common/machine.go:
```diff
@@ -0,0 +1,245 @@
+
+// machineIndex returns the index of the machine, the numeric suffix of the machine name.
+func machineIndex(machine machinev1beta1.Machine) (int, error) {
+	nameSuffix := machine.ObjectMeta.Name[len(machine.ObjectMeta.Name)-1]
```
Hmm interesting, given we are planning to run these tests in a controlled environment, I wouldn't expect to see anything other than 0, 1, 2 for our tests here, which is why last index should be ok. I can look at updating it to grab a number using a regex though 🤔 Shouldn't be too hard

JoelSpeed commented on test/e2e/framework/framework.go:
```diff
@@ -191,3 +236,75 @@ func getPlatformSupportLevel(k8sClient runtimeclient.Client) (PlatformSupportLev
+
+// nextAWSInstanceSize returns the next AWS instance size in the series.
+// In AWS terms this normally means doubling the size of the underlying instance.
+// For example:
+// - m6i.large -> m6i.xlarge
+// - m6i.xlarge -> m6i.2xlarge
+// - m6i.2xlarge -> m6i.4xlarge
+// This should mean we do not need to update this when the installer changes the default instance size.
+func nextAWSInstanceSize(current string) (string, error) {
+	// Regex to match the AWS instance type string.
+	re := regexp.MustCompile(`(?P<family>[a-z0-9]+)\.(?P<multiplier>\d)?(?P<size>[a-z]+)`)
+
+	values := re.FindStringSubmatch(current)
+	if len(values) == 0 || len(values) != 4 {
+		return "", fmt.Errorf("%w: %s", errInstanceTypeUnsupportedFormat, current)
+	}
+
+	family := values[1]
+	size := values[3]
+
+	if multiplier := values[2]; multiplier == "" {
+		switch size {
+		case "large":
+			return fmt.Sprintf("%s.xlarge", family), nil
+		case "xlarge":
+			return fmt.Sprintf("%s.2xlarge", family), nil
+		default:
+			return "", fmt.Errorf("%w: %s", errInstanceTypeNotSupported, current)
+		}
+	}
+
+	multiplierInt, err := strconv.Atoi(values[2])
+	if err != nil {
+		// This is a panic because the multiplier should always be a number.
+		panic("failed to convert multiplier to int")
+	}
+
+	return fmt.Sprintf("%s.%d%s", family, multiplierInt*2, size), nil
+}
+
```
I'll add a unit test that verifies the instance sizes in and out

damdo replied:
Thanks

JoelSpeed commented on test/e2e/framework/utils.go:
```diff
@@ -17,17 +17,73 @@ limitations under the License.
+	}
+}
+
+// ControlPlaneMachineSetSelectorLabels are the set of labels use to select
+// control plane machines within the cluster.
+func ControlPlaneMachineSetSelectorLabels() map[string]string {
+	return map[string]string{
+		"machine.openshift.io/cluster-api-machine-role": "master",
+		"machine.openshift.io/cluster-api-machine-type": "master",
+	}
+}
+
```
At the moment, openshift-install, AFAIK creates clusters only with `master`, I think if they are now creating them with `control-plane`, we can switch this to `control-plane`, but I think, because this is for E2E on fresh clusters, we don't need to support both simultaneously right? We support them simultaneously for upgrades sake and to support the legacy values

damdo replied:
Yes makes sense, the important thing here is to remember to change these when we'll migrate the installer.

damdo reviewed (commented):
Changes looks good. Thanks for addressing them.

damdo reviewed (commented):
/lgtm

JoelSpeed commented:
/retest
